docs/src/writing-constructs.rst

```diff
@@ -0,0 +1,169 @@
+  **Subscription** but also setting appropriate IAM permissions), make
+  both things happen in a single call (to **subscribeQueue()**). The user
+  should be mentally thinking about the result they're trying to achieve,
+  not the various steps necessary to get there.
+* Make sure you are not introducing new concepts or terminology. For example
+  don't be temped to replace SQS's **dataKeyReusePeriod** with **keyRotation**
+  because it will be hard for people to diagnose problems. They won't be able to
+  just search for *sqs dataKeyReuse* and find topics on it.
+* Indicate units of measurement in property names that don't use a strong type.
+  Use **milli**, **sec**, **min**, **hr**, **Bytes**, **KiB**, **MiB**, **GiB**,
+  etc.
+* Optimize for the common case; example: **Fleet** accepts a **VPC** and deploys
+  in the private subnet by default because that's the common case, but has an
+  option to **placementOptions** for the special case.
+* Use Typescript; in order for construct libraries to be reusable across
+  programming languages (using jsii), they need to be authored in a language
+  that can compile to a jsii assembly. At the moment, the only supported
+  language is TypeScript.
+
+Implementation Details
+======================
+
+* Every construct consist of an exported class (**MyConstruct**) and an exported
+  interface (**MyConstructProps**) that defines the parameters for these
+  classes.
+* Constructs are (mostly) immutable; sometimes we use mutating methods if it
+  makes sense or reads better (for example, for event handlers we will write
+  `builder.onBuild(topic)`), but configuration props are usually passed into
+  the constructor, after which they cannot be changed anymore.
+* Most of the logic happens in the constructor; the constructor will build up
+  the state of the construct (what children it has, which ones are always
+  there and which ones are optional, etc.).
+* Validate as early as possible; throw an `Error` in the constructor if the
+  parameters don't make sense. Only if you want to validate mutations that can
+  occur after construction time, override the `validate()` method.
+* Avoid unneeded hierarchy in props; try to keep the props interface flat to
+  help discoverability.
+* Don't expose **Token** s to your consumers; tokens are an implementation
+  mechanism for one of two purpose: representing CloudFormation intrinsic
+  values, or representing lazily evaluated values. They can be used for
+  implementation purposes, but use more specific types as part of your public
+  API.
+* Don't use type unions (**string | BucketName**); they do not translate well
+  across **jsii**. Instead, either just accept a **string** or a **BucketName**.
+  One can always be translated into the other (either by Token stringification
+  or by constructing a **BucketName** from a literal).
+* Accept objects instead of ARNs; when accepting other resources as parameters,
```

Comment:
This probably applies for names in a number of scenarios. I think places where you'd want to accept a `string` in that happens to need to be - to re-use the example - a Bucket Name, you should probably get a BucketRef instead... So at least the compiler tells you this is a bucket-related thing.